File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.cpp
class Solution {
public:
    unordered_map<int,vector<int>>mp;
    vector<int>ans;
    void dfs(int node,int prev){
        ans.push_back(node);
        for(int neighbor:mp[node]){
            if(neighbor!=prev){
                dfs(neighbor,node);
            }
        }
    }
    
    vector<int> restoreArray(vector<vector<int>>& adjacentPairs) {
        int n=adjacentPairs.size();
        for(int i=0;i<n;i++){
            mp[adjacentPairs[i][0]].push_back(adjacentPairs[i][1]);
            mp[adjacentPairs[i][1]].push_back(adjacentPairs[i][0]);
        }
        int root=0;
        for(auto it:mp){
            if(it.second.size()==1){
           root=it.first;
           break;
        }    
    
    }
        dfs(root,INT_MAX);
        return ans;
            
    }
};

File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.java
class Solution {
    Map<Integer, List<Integer>>mp=new HashMap<>();
    int []ans;
    void dfs(int node,int prev,int i){
        ans[i]=node;
        for (int neighbor : mp.get(node)) {
            if (neighbor != prev) {
                dfs(neighbor, node,i+1);
            }
        }
    }
    public int[] restoreArray(int[][] adjacentPairs) {
        int n = adjacentPairs.length;
        for (int[] edge : adjacentPairs) {
            int x = edge[0];
            int y = edge[1];

            if (!mp.containsKey(x)) {
                mp.put(x, new ArrayList());
            }

            if (!mp.containsKey(y)) {
                mp.put(y, new ArrayList());
            }

            mp.get(x).add(y);
            mp.get(y).add(x);
        }
        ans=new int[mp.size()];
        int root = 0;
        for (int it:mp.keySet()) {
            if (mp.get(it).size() == 1) {
                root = it;
            }
        }

        dfs(root, Integer.MAX_VALUE,0);
        return ans;
 }

}